File: 백준/Gold/11657. 타임머신/타임머신.py
INF = float('inf')

def BellmanFord(graph, start):
    size = len(graph)
    dist = [INF] * (size)

    # 시작점에서의 비용은 0으로 설정
    dist[start] = 0

    # 정점의 개수만큼 반복
    for _ in range(size - 1):
        for u in range(1, size):
            for v, w in graph[u]:
                if dist[u] != INF and dist[u] + w < dist[v]:
                    dist[v] = dist[u] + w

    # 음수 사이클 탐지
    for u in range(1, size):
        for v, w in graph[u]:
            if dist[u] != INF and dist[u] + w < dist[v]:
                # 음수 가중치 사이클 발견되는 경우
                return None
            
    return dist[1:]


n, m = map(int, input().split())

graph = [[] for _ in range(n + 1)]

for _ in range(m):
    start, end, cost = map(int, input().split())
    graph[start].append((end, cost))

dist_result = BellmanFord(graph, 1)

if dist_result is not None:
    for i in range(1, len(dist_result)):
        print(dist_result[i] if dist_result[i] != INF else -1)
else:
    print(-1)

File: 백준/Gold/11657. 타임머신/타임머신.cc
#include <bits/stdc++.h>
using namespace std;

typedef long long ll;

int n, m, src, tgt, cost;
const ll INF = 1e10;

pair<bool, vector<ll>> bellman_ford(const vector<vector<pair<ll, ll>>> & graph, ll start_node) {
	bool is_negative_cycle = false;
	
	ll n = graph.size();

	vector<ll> dist(n, INF);
	dist[start_node] = 0;

	for (ll k = 1; k < n; ++k) {
		for (ll u = 1; u < n; ++u) {
			for (pair<ll, ll> p : graph[u]) {
				ll v = p.first, cost = p.second;
				if (dist[u] != INF && dist[u] + cost < dist[v]) {
					dist[v] = dist[u] + cost;
				}
			}
		}
	}

	for (ll u = 1; u < n; ++u) {
		for (pair<ll, ll> p : graph[u]) {
			ll v = p.first, cost = p.second;
			if (dist[u] != INF && dist[u] + cost < dist[v]) {
				is_negative_cycle = true;
				break;
			}
		}

		if (is_negative_cycle) break;
	}

	// 첫번째 요소 삭제
	dist = vector<ll> (dist.begin() + 1, dist.end());

	return pair<bool, vector<ll>> (is_negative_cycle, dist);
}

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL); cout.tie(NULL);

	cin >> n >> m;

	vector<vector<pair<ll, ll>>> graph(n + 1);

	for (int i = 0; i < m; ++i) {
		cin >> src >> tgt >> cost;
		graph[src].push_back({tgt, cost});
	}

	bool is_cycle;
	vector<ll> dist_result;

	tie(is_cycle, dist_result) = bellman_ford(graph, 1);

	if (is_cycle) {
		cout << -1;
	}
	else {
		for (int i = 1; i < n; ++i) {
			if (dist_result[i] == INF) {
				cout << -1 << "\n";
			}
			else {
				cout << dist_result[i] << "\n";
			}
		}
	}

	return 0;
}